### PR TITLE
feat: Check role extraction and add role to user

### PR DESCRIPTION
**Pull request by gsandok**
Adapt the current Dex setup to correctly extract the role based on the cookie given
allows multiple types of role extraction based on groups and email

REV: DSN-RQTDRW

### CODE REVIEW

**Review by jdvgh — COMMENTED** [review]
In general it looks good to me.
Just added some comments regarding the claim handling

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 1]
```diff
@@ -133,13 +149,27 @@ func DexLoginInterceptor(
...
+			httpCtx = checkPolicy(httpCtx, w, req, groupName)
+		}
+	case claims["sub"].(string) != "":
+		httpCtx = checkPolicy(httpCtx, w, req, claims["sub"].(string))
+	default:
+		http.Error(w, "unable to parse token with expected fields for DEX login", http.StatusBadRequest)
+		return
+	}
+
+	httpCtx = context.WithValue(httpCtx, "claims", claims)
```
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:172]: Yes I need to check that it works as expected but this is adding claims to the header !

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 2]
```diff
@@ -123,6 +124,21 @@ func GoogleIAPInterceptor(
...
 
+func checkPolicy(httpCtx context.Context, w http.ResponseWriter, req *http.Request, userGroup string) context.Context {
+	policy, err := auth.ReadRbacPolicy(true, "/etc/policy.csv")
+	if err != nil {
+		http.Error(w, "unable to access RBAC policy to validate login", http.StatusBadRequest)
+		return nil
+	}
+	for _, policyGroup := range policy.Groups {
+		if policyGroup.Group == userGroup {
+			auth.WriteUserRoleToHttpHeader(req, policyGroup.Role)
```
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:135]: I am not sure because currently we only expect one so for now I keep over writing but I am unsure if I should change the role from a string into a list

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 3]
```diff
@@ -123,6 +124,21 @@ func GoogleIAPInterceptor(
 	httpHandler(w, req)
 }
 
+func checkPolicy(httpCtx context.Context, w http.ResponseWriter, req *http.Request, userGroup string) context.Context {
+	policy, err := auth.ReadRbacPolicy(true, "/etc/policy.csv")
```
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:128]: Currently this file path is correct since it is what we use on the cd service but i will change it to a variable if we agree with this approach

File: pkg/auth/auth.go [hunk 4]
```diff
@@ -81,6 +82,12 @@ func WriteUserRoleToGrpcContext(ctx context.Context, userRole string) context.Co
 	return metadata.AppendToOutgoingContext(ctx, HeaderUserRole, Encode64(userRole))
 }
 
+// WriteUserRoleToGrpcContext adds the user claims to the GRPC context.
+// Only used when RBAC is enabled.
+func WriteUserClaimsToGrpcContext(ctx context.Context, claims string) context.Context {
```
[the remark this answers is not in the document]
Comment (gsandok) [reply pkg/auth/auth.go:87]: Yeah I am not calling it and should remove it since I was unable to get it to work with jwt.MapClaims

File: pkg/auth/dex.go [hunk 5]
```diff
@@ -271,39 +270,43 @@ func (a *DexAppClient) oauth2Config(scopes []string) (c *oauth2.Config, err erro
 	}, nil
 }
 
+type MapClaims struct {
```
[the remark this answers is not in the document]
Comment (gsandok) [reply pkg/auth/dex.go:273]: Yeah was jsut using this as a place holder to organize the variables i would need, can delete now !

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 6]
```diff
@@ -133,13 +149,27 @@ func DexLoginInterceptor(
...
+	claims, err := auth.VerifyToken(req.Context(), req, clientID, baseURL)
 	if err != nil {
 		// If user is not authenticated redirect to the login page.
 		http.Redirect(w, req, auth.LoginPATH, http.StatusFound)
 	}
-	auth.WriteUserRoleToHttpHeader(req, role)
-	httpCtx := auth.WriteUserRoleToGrpcContext(req.Context(), role)
+	var httpCtx context.Context
+
+	switch {
```
[the remark this answers is not in the document]
Comment (jdvgh) [reply services/frontend-service/pkg/interceptors/interceptors.go:159]: Straightup copy:

```
sub is the subject of the JWT– the user who requested the token, typically an email address. By using it, we can send and receive a small, unique piece of data about a user along with each request, without any additional requests.
```

https://jakeworth.com/til-what-is-the-jwt-sub/

So I think it might be the mail or also a token so it is probably needed to identify the user and important to be passed to the cd-service

File: services/frontend-service/pkg/interceptors/interceptors.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdvgh) [reply services/frontend-service/pkg/interceptors/interceptors.go:135]: The JWT-Claim of the user could contain multiple groups  so it could also contain multiple roles which are assigned to the different groups is what I was thinking of.

But maybe this is just an edge case which will not happen 👍

File: services/frontend-service/pkg/interceptors/interceptors.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:159]: But then it seems that nothing from the screenshot you provided in your last story can be used for assigning the group for getting the role?

File: services/frontend-service/pkg/interceptors/interceptors.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:135]: I feel that that is a bigger issue that we should maybe decide, since some roles could have differing policies, but that can be expanded to be allowed later?

File: services/frontend-service/pkg/interceptors/interceptors.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jdvgh) [reply services/frontend-service/pkg/interceptors/interceptors.go:159]: Another remark I just noted (cause my Dex Setup failed because of it):
You might need to catch the error of the cast `([]interface{})`. I got a panic e.g. at when "groups" is not present in the claim.

[the remark this answers is not in the document]
Comment (jdvgh) [reply services/frontend-service/pkg/interceptors/interceptors.go:159]: Just for reference the stack-trace - The errror was the following:
```log
2024/04/29 19:59:46 http: panic serving 127.0.0.6:42125: interface conversion: interface {} is []string, not []interface {}
goroutine 495 [running]:
net/http.(*conn).serve.func1()
    /usr/local/go/src/net/http/server.go:1868 +0xb9
panic({0xd30a20?, 0xc00023b2c0?})
    /usr/local/go/src/runtime/panic.go:920 +0x270
github.com/freiheit-com/kuberpult/pkg/auth.VerifyToken({0x100b920, 0xc00023a7b0}, 0x1e000000001?, {0xc000046708, 0xd}, {0xc00004a067, 0x2f})
    /kp/pkg/auth/dex.go:300 +0x48e
github.com/freiheit-com/kuberpult/services/frontend-service/pkg/interceptors.DexLoginInterceptor({0x1009f38, 0xc00059c380}, 0xc000323900, 0xc000305288, {0xc000046708, 0xd}, {0xc00004a067, 0x2f})
    /kp/services/frontend-service/pkg/interceptors/interceptors.go:137 +0xa5
github.com/freiheit-com/kuberpult/services/frontend-service/pkg/cmd.runServer.func3({0x1009f38?, 0xc00059c380?}, 0x78aa08d9e478?)
    /kp/services/frontend-service/pkg/cmd/server.go:331 +0x126
net/http.HandlerFunc.ServeHTTP(0xc000067400?, {0x1009f38?, 0xc00059c380?}, 0xc000679520?)
    /usr/local/go/src/net/http/server.go:2136 +0x29
net/http.(*ServeMux).ServeHTTP(0xdd1a40?, {0x1009f38, 0xc00059c380}, 0xc000323900)
    /usr/local/go/src/net/http/server.go:2514 +0x142
github.com/freiheit-com/kuberpult/services/frontend-service/pkg/cmd.runServer.func5({0x1009f38, 0xc00059c380}, 0xc000323900)
    /kp/services/frontend-service/pkg/cmd/server.go:429 +0x98c
net/http.HandlerFunc.ServeHTTP(0x100b920?, {0x1009f38?, 0xc00059c380?}, 0xb?)
    /usr/local/go/src/net/http/server.go:2136 +0x29
github.com/freiheit-com/kuberpult/services/frontend-service/pkg/cmd.(*Auth).ServeHTTP.func1({0x100b920?, 0xc00023a600?})
    /kp/services/frontend-service/pkg/cmd/server.go:538 +0x579
github.com/freiheit-com/kuberpult/pkg/logger.Wrap({0x100b958, 0xc00060a1e0}, 0xc000679990)
    /kp/pkg/logger/logger.go:77 +0x283
github.com/freiheit-com/kuberpult/services/frontend-service/pkg/cmd.(*Auth).ServeHTTP(0xc00023a570?, {0x1009f38?, 0xc00059c380?}, 0x78a9c22087c0?)
    /kp/services/frontend-service/pkg/cmd/server.go:512 +0x8c
github.com/freiheit-com/kuberpult/pkg/setup.(*CORSMiddleware).ServeHTTP(0xc0000110f8, {0x1009f38, 0xc00059c380}, 0xc000322d00)
    /kp/pkg/setup/cors.go:44 +0x933
net/http.(*ServeMux).ServeHTTP(0x169d960?, {0x1009f38, 0xc00059c380}, 0xc000322d00)
    /usr/local/go/src/net/http/server.go:2514 +0x142
net/http.serverHandler.ServeHTTP({0xc000384720?}, {0x1009f38?, 0xc00059c380?}, 0x6?)
    /usr/local/go/src/net/http/server.go:2938 +0x8e
net/http.(*conn).serve(0xc000098b40, {0x100b920, 0xc000528120})
    /usr/local/go/src/net/http/server.go:2009 +0x5f4
created by net/http.(*Server).Serve in goroutine 12
    /usr/local/go/src/net/http/server.go:3086 +0x5cb

```

[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:159]: Already removed that line here

File: docker-compose-earthly.yml [hunk 7]
```diff
@@ -41,6 +41,7 @@ services:
       - KUBERPULT_GIT_BRANCH=main
       - KUBERPULT_API_ENABLE_DESPITE_NO_AUTH=false
       - KUBERPULT_BATCH_CLIENT_TIMEOUT=2m
+      - KUBERPULT_DEX_RBAC_POLICY_PATH=/etc/policy.csv
```
Comment (jdvgh) [inline docker-compose-earthly.yml:44]: this is not mounted to the frontend in the kuberpult chart yet, just so we do not forget!

File: charts/kuberpult/templates/frontend-service.yaml [hunk 8]
```diff
@@ -153,6 +153,8 @@ spec:
           value: "{{ .Values.auth.dexAuth.baseURL }}"
         - name: KUBERPULT_DEX_SCOPES
           value: "{{ .Values.auth.dexAuth.scopes }}"
+        - name: KUBERPULT_DEX_RBAC_POLICY_PATH
```
Comment (sven-urbanski-freiheit-com) [inline charts/kuberpult/templates/frontend-service.yaml:156]: ok, I am just surprised we didn't have that already :D

File: charts/kuberpult/templates/frontend-service.yaml [hunk 9]
```diff
@@ -183,12 +185,21 @@ spec:
         - name: keyring
           mountPath: /keyring
 {{- end }}
```
Comment (sven-urbanski-freiheit-com) [inline charts/kuberpult/templates/frontend-service.yaml:187]: please adda test (at least for the new env var) in chart_test.go

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 10]
```diff
@@ -132,16 +149,44 @@ func DexLoginInterceptor(
...
+		}
+	case []string:
+		httpCtx = CheckPolicy(httpCtx, w, req, strings.Join(val, ","), DexRbacPolicyPath)
+	case string:
+		httpCtx = CheckPolicy(httpCtx, w, req, val, DexRbacPolicyPath)
+	}
+
+	if claims["email"].(string) != "" {
+		httpCtx = CheckPolicy(httpCtx, w, req, claims["email"].(string), DexRbacPolicyPath)
+	}
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/pkg/interceptors/interceptors.go:186]: please rephrase this so that we have the structure 
```
if errorCase {
   return err
}
if otherErrCase {
return otherErr
}
return success.
```
Otherwise it's hard to follow what is happening, and it saves you from checking the claims["email"] twice.

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 11]
```diff
@@ -124,6 +125,22 @@ func GoogleIAPInterceptor(
 	httpHandler(w, req)
 }
 
+func CheckPolicy(httpCtx context.Context, w http.ResponseWriter, req *http.Request, userGroup, DexRbacPolicyPath string) context.Context {
+	policy, err := auth.ReadRbacPolicy(true, DexRbacPolicyPath)
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/pkg/interceptors/interceptors.go:129]: the parsing (in `ReadRbacPolicy`) should happen on startup, not on a request basis.
 And if there's an error in it, we need to log the error.

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 12]
```diff
@@ -124,6 +125,22 @@ func GoogleIAPInterceptor(
...
+	if err != nil {
+		http.Error(w, "unable to access RBAC policy to validate login", http.StatusBadRequest)
+		return nil
+	}
+	for _, policyGroup := range policy.Groups {
+		if policyGroup.Group == userGroup {
+			auth.WriteUserRoleToHttpHeader(req, policyGroup.Role)
+			httpCtx = auth.WriteUserRoleToGrpcContext(req.Context(), policyGroup.Role)
+			httpCtx = context.WithValue(httpCtx, auth.HeaderUserRole, auth.Encode64(policyGroup.Role)) //nolint: go-staticcheck
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/pkg/interceptors/interceptors.go:138]: this is a bit hard to follow.
Do we really need to write this role into 3 places? Seems a bit excessive.
Please check if this is really necessary - and if it is, please add a comment, e.g
`auth.WriteUserRoleToHttpHeader(req, policyGroup.Role) // this is needed when service foo calls service bla over http`
etc.

**Review by sven-urbanski-freiheit-com — CHANGES_REQUESTED** [review]
.

File: services/frontend-service/pkg/interceptors/interceptors.go [hunk 13]
```diff
@@ -132,16 +149,44 @@ func DexLoginInterceptor(
...
+		for _, group := range val {
+			groupName := strings.Trim(group.(string), "\"")
+			httpCtx = CheckPolicy(httpCtx, w, req, groupName, DexRbacPolicyPath)
+		}
+	case []string:
+		httpCtx = CheckPolicy(httpCtx, w, req, strings.Join(val, ","), DexRbacPolicyPath)
+	case string:
+		httpCtx = CheckPolicy(httpCtx, w, req, val, DexRbacPolicyPath)
+	}
+
```
[the remark this answers is not in the document]
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go:183]: Will do, this was to handle all formats that the groups could be sent it from different types of Claims

File: services/frontend-service/pkg/interceptors/interceptors.go (hunk 11 above)
Comment (gsandok) [reply services/frontend-service/pkg/interceptors/interceptors.go]: Adapted it to be read only once

File: services/frontend-service/pkg/cmd/server.go [hunk 14]
```diff
@@ -327,13 +332,14 @@ func runServer(ctx context.Context) error {
...
 		if c.DexEnabled {
-			interceptors.DexLoginInterceptor(w, req, httpHandler.Handle, c.DexClientId, c.DexBaseURL)
+			interceptors.DexLoginInterceptor(w, req, httpHandler.Handle, c.DexClientId, c.DexBaseURL, policy)
 			return
 		}
 		httpHandler.Handle(w, req)
 	})
 	for _, endpoint := range []string{
 		"/environments",
+		"/api.v1.BatchService/ProcessBatch",
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/pkg/cmd/server.go:342]: this does not look like a rest endpoint. Is this line correct?

File: services/frontend-service/pkg/cmd/server.go [hunk 15]
```diff
@@ -542,6 +564,24 @@ func (p *Auth) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+func getUserFromDex(w http.ResponseWriter, req *http.Request, clientID, baseURL, DexRbacPolicyPath string) *auth.DexAuthContext {
+	policy, err := auth.ReadRbacPolicy(c.DexEnabled, c.DexRbacPolicyPath)
```
Comment (sven-urbanski-freiheit-com) [inline services/frontend-service/pkg/cmd/server.go:568]: we already did `auth.ReadRbacPolicy` during startup. We shouldn't have to do it again here.

Comment (gsandok) [reply services/frontend-service/pkg/cmd/server.go]: Agreed!

**Review by sven-urbanski-freiheit-com — APPROVED** [review]
